wallet/src/de/schildbach/wallet/ui/pop/PopActivity.java

```diff
@@ -0,0 +1,446 @@
+import java.net.URL;
+import java.util.List;
+
+/**
+ * @author Kalle Rosenbaum
+ */
+public class PopActivity extends AbstractWalletActivity
+{
+	private static final Logger log = LoggerFactory.getLogger(PopActivity.class);
+
+	public static final String INTENT_EXTRA_POP_INTENT = "pop_intent";
+
+	private PopRequestURI popRequestURI;
+	private Transaction transactionToProve;
+	private Button proveButton;
+	private State currentState = null;
+
+	private enum State
+	{
+		INPUT(R.string.pop_send_pop), // asks for confirmation
+		DECRYPTING(R.string.send_coins_fragment_state_decrypting),
+		SIGNING(R.string.send_coins_preparation_msg),
+		SENDING(R.string.send_coins_sending_msg),
+		SENT(R.string.pop_sent),
+		SUCCESS(R.string.pop_proven),
+		FAILED(R.string.send_coins_failed_msg); // sending
+																																	// states
+		private final int resId;
+
+		State(final int resId)
+		{
+			this.resId = resId;
+		}
+
+		private int getResId()
+		{
+			return resId;
+		}
+	}
+
+	public static void start(final Context context, final PopIntent popIntent)
+	{
+		final Intent intent = new Intent(context, PopActivity.class);
+		intent.putExtra(INTENT_EXTRA_POP_INTENT, popIntent);
+		context.startActivity(intent);
+	}
+
+	@Override
+	protected void onCreate(final Bundle savedInstanceState)
+	{
+		super.onCreate(savedInstanceState);
+
+		setContentView(R.layout.pop_content);
+
+		// We can come here either from an intent filter in AndroidManifest.xml or from
+		// WalletActivity.onActivityResult after scanning
+		// If from intent filter, we have a Uri. If from WalletActivity, we have a PopIntent.
+		PopIntent popIntent = getIntent().getParcelableExtra(INTENT_EXTRA_POP_INTENT);
+		if (popIntent == null)
+		{
+			// We are started from the intent filter
+			String input = getIntent().getData().toString();
+			try
+			{
+				popIntent = PopIntent.fromPopRequestURI(new PopRequestURI(input));
+			}
+			catch (IllegalArgumentException e)
+			{
+				log.info("Got invalid btcpop uri: '" + input + "'", e);
+				resultMessage(false, R.string.pop_input_parser_invalid_btcpop_uri, input);
+				finish();
+				return;
+			}
+		}
+		popRequestURI = popIntent.getPopRequestURI();
+
+		WalletApplication walletApplication = getWalletApplication();
+		Wallet wallet = walletApplication.getWallet();
+
+		List<Transaction> transactions = wallet.getTransactionsByTime();
+		for (Transaction transaction : transactions)
+		{
+			if (popRequestURI.getTxid() != null && !transaction.getHash().equals(popRequestURI.getTxid()))
+			{
+				continue;
+			}
+			Coin value = transaction.getValue(wallet);
+			if (popRequestURI.getAmountSatoshis() != null && value.longValue() != popRequestURI.getAmountSatoshis())
+			{
+				continue;
+			}
+			if (popRequestURI.getLabel() != null && !popRequestURI.getLabel().equals(transaction.getMemo()))
+			{
+				continue;
+			}
+			transactionToProve = transaction;
+			break;
+		}
+
+		if (transactionToProve == null)
```

Comment:
I guess this section is not meant to stay, right? A longToast is still much to short to read all that stuff, and it won't mean much to the user anyway.